drivers: ieee802154: Add RX improvements

nandojve opened the pull request:
The current version of at86rf2xx RX implementation don't uses advanced capabilities offered by the transceiver. This introduces SRAM access space to gatter PHR information in parallel with transceiver frame reception. It allows improve RX reception by handling properlly the frame protection feature
removing transceiver states changes.

This change can improve RX response time significantly but will add a small delay at TX to make sure that transceiver isn't finishing RX ACK process on advanced reception mode. On the other hand, if we need a TX just after RX, this will guarantee that TX will be set ASAP after finish RX ACK process.

This was fully tested on HW and ready for merge.

nandojve committed:
drivers: ieee802154: rf2xx: Add SRAM read method

Add SRAM read method to enable advanced uses on at86rf2xx driver.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: ieee802154: rf2xx: Add RX improvements

The current version of at86rf2xx RX implementation don't uses advanced
capabilities offer by the transceiver. This access SRAM space to gatter
PHR information in parallel with transceiver frame reception. It allows
improve RX reception by handling properlly the frame protection feature
removing transceiver states changes.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

rlubos commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -133,6 +152,13 @@ static void rf2xx_trx_rx(struct device *dev)
+	// NEED CHECK HERE IF EVERYTHING ABOVE ARE OK BEFORE DROP rf2xx_iface_frame_read
+	//pkt_len = ctx->rx_phr;
+	if (pkt_len != ctx->rx_phr)
+	{
```
The opening bracket should be moved to the line above.

rlubos commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -133,6 +152,13 @@ static void rf2xx_trx_rx(struct device *dev)
+	//pkt_len = ctx->rx_phr;
+	if (pkt_len != ctx->rx_phr)
+	{
+		LOG_INF("pkt_len != ctx->rx_pkt_len");
```
The comment doesn't match the code, it seems that `rx_pkt_len` was renamed to `rx_phr`?

rlubos commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -230,26 +256,17 @@ static void rf2xx_thread_main(void *arg)
+			 */
+			rf2xx_iface_sram_read(dev, 0, &ctx->rx_phr, 1);
+		}
+		// IF FOR SOME REASON RX_START and TRX_END WE CAN PROCESS BOTH
```
`/* */` please.

rlubos commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -230,26 +256,17 @@ static void rf2xx_thread_main(void *arg)
 			k_timer_start(&ctx->trx_isr_timeout, K_MSEC(10), 0);
-		} else if (isr_status & (1 << RF2XX_TRX_END)) {
+
+			/* get PHR len and store at ctx->rx_pkt_len
```
It's `ctx->rx_phr`.

rlubos commented on drivers/ieee802154/ieee802154_rf2xx_iface.c:
```diff
@@ -248,3 +248,42 @@ void rf2xx_iface_frame_write(struct device *dev,
 }
+
+void rf2xx_iface_sram_read(struct device *dev,
+				  u8_t address,
```
The alignment seems a bit off here and in the function declaration.

nandojve replied:
Yes, this is annoying. I can't find a tool that I can do it on any GUI to programmatically fix that. I'm using VS-Code that show one thing, git diff shows another thing, gedit show another thing and github shows another things too and for sure no one of that tools are right.

I'll re-check is chackpath is enable on git commit.

rlubos reviewed (commented):
The changes look reasonable, I'm not familiar with the actual HW to provide more valuable feedback I'm afraid. I've left comments on some nits spotted.

tbursztyka commented:
Seems fine to me. Once you've tested it I'll +1

nandojve committed:
boards: arm: atsamr21_xpro: Fix corrupted frames on rf2xx

When using internal IEEE 802.15.4 transceiver AT86RF233 at SPI speeds
above 6MHz was detected that frame buffer read returns corrupted data.
This set spi-max-frequency to 6MHz at atsamr21_xpro.dts to ensure best
relation of performance vs reliability.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
At below you can find a wireshark print. The `samples/net/socktets/echo_[client|server]` applications is working prety well.

![image](https://user-images.githubusercontent.com/7027887/74094944-0837a300-4ac8-11ea-8549-1896dfe92293.png)

I faced problems to run with atsamr21 SoC because SPI speed was to high. After changed it to 6MHz it started to work. This take me majority of tests time and was not easy to realise that SPI speed corrupted frame read. The idea to maintain frame read is to keep auto frame buffer protection and with one SPI transaction read all data payload, LQI, ED and status.

nandojve committed:
drivers: ieee802154: rf2xx: Refactor rf2xx_thread_main

Current rf2xx_thread_main code have too many if/for/while imbrication.
Extract methods from rx2xx_thread_main for better readability.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
superseded by #23242
